### PR TITLE
Fix bug The helper "dialog" is not defined (Undefined DialogHelper Class)

### LINKED ISSUE #4327: The helper "dialog" is not defined (Undefined DialogHelper Class)
#### Sonata packages

```bash
$ composer show sonata-project/*
sonata-project/admin-bundle              3.13.0 The missing Symfony Admin Generator
sonata-project/block-bundle              3.3.0  Symfony SonataBlockBundle
sonata-project/cache                     1.0.7  Cache library
sonata-project/core-bundle               3.2.0  Symfony SonataCoreBundle
sonata-project/doctrine-orm-admin-bundle 3.1.3  Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/exporter                  1.7.1  Lightweight Exporter library
```

#### Symfony packages

```bash
$ composer show symfony/*
symfony/monolog-bundle     v3.0.3 Symfony MonologBundle
symfony/phpunit-bridge     v3.2.3 Symfony PHPUnit Bridge
symfony/polyfill-apcu      v1.3.0 Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-intl-icu  v1.3.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring  v1.3.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php56     v1.3.0 Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70     v1.3.0 Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util      v1.3.0 Symfony utilities for portability of PHP codes
symfony/security-acl       v3.0.0 Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle v2.4.2 Symfony SwiftmailerBundle
symfony/symfony            v3.2.3 The Symfony PHP framework
```

#### PHP version

```bash
$ php -v
PHP 7.1.1-1+deb.sury.org~trusty+1 (cli) (built: Jan 20 2017 09:43:29) ( NTS )
```

## Subject

Undefined DialogHelper Class:

**\Sonata\AdminBundle\Command\GenerateObjectAclCommand::execute**
```
    /**
     * {@inheritdoc}
     */
    public function execute(InputInterface $input, OutputInterface $output)
    {
        /** @var $dialog \Symfony\Component\Console\Helper\DialogHelper */
        $dialog = $this->getHelperSet()->get('dialog');
```

## Steps to reproduce
`$ php bin/console sonata:admin:generate-object-acl `

## Expected results
All happy

## Actual results

```
  [Symfony\Component\Console\Exception\InvalidArgumentException]
  The helper "dialog" is not defined.
```

### PR DESCRIPTION
**Pull request by sinoptik**
I am targetting this branch, because I found bug The helper "dialog" is not defined (Undefined DialogHelper Class)

Closes #4327

## Changelog

```markdown
### Fixed
- Fixed the bug that caused an error "The helper "dialog" is not defined." on Symfony3 with new `\Sensio\Bundle\GeneratorBundle\Command\Helper\QuestionHelper` when you run command "sonata:admin:generate-object-acl". 
```

## To do
Delete or re-right Command/QuestionableCommand.php

https://github.com/sonata-project/SonataAdminBundle/issues/4327

### CODE REVIEW

**Comment by jordisala1991** [comment]
@sinoptik can you fill the PR template correctly?

**Comment by vedro-compota** [comment]
@jordisala1991 this's important fix -- please approve it without waitnig for "templates" )

File: Command/QuestionableCommand.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            $questionHelper = $this->getHelper('question');
+
+            if (!$questionHelper instanceof QuestionHelper) {
+                $questionHelper = new QuestionHelper();
+                $this->getHelperSet()->set($questionHelper);
+            }
+        }
+
+        return $questionHelper;
+    }
```
Comment (greg0ire) [inline Command/QuestionableCommand.php]: Please make all methods final

**Comment by greg0ire** [comment]
@vedro-compota please do not disregard our policies. If you want to speed things up, you can carry the PR yourself.

**Comment by vedro-compota** [comment]
> @vedro-compota please do not disregard our policies. If you want to speed things up, you can carry the PR yourself.

@greg0ire should I fork  @sinoptik 's fork? or there is another way to help?

**Comment by OskarStark** [comment]
> @greg0ire should I fork @sinoptik 's fork? or there is another way to help?

I would say, fork the adminbundle and create a new PR

**Comment by greg0ire** [comment]
You can do both. I'm on my phone, but @Oskarstark can find instructions to do this by clicking "merge manually" and show them to you

**Comment by sinoptik** [comment]
Sorry for long response.
I updated the pull request message
Make all methods final

**Comment by OskarStark** [comment]
I updated the changelog a bit, @greg0ire what do you think or still too much?

**Comment by greg0ire** [comment]
The changelog should be something that the users can read, and understand the benefit of this PR. Right now I don't understand it.

**Comment by sinoptik** [comment]
@greg0ire What you can say about current changelog?

**Comment by greg0ire** [comment]
> - `QuestionableCommand` to handle both deprecated `\Symfony\Component\Console\Helper
\DialogHelper` and new `\Sensio\Bundle\GeneratorBundle\Command\Helper\QuestionHelper`

What does it fix for me, a sonata bundle user?

> - `GenerateObjectAclCommand`and `GenerateAdminCommand ` now extends from new `QuestionableCommand` for work with console askAndValidate, askConfirmation method

I am a sonata admin bundle user, should I care about that? What does it change for me?

> - Fixed the bug that caused an error "The helper "dialog" is not defined." on Symphony3 when you run command "sonata:admin:generate-object-acl"

That is the good part of the changelog. As a user, I understand what it will change for me.

**Comment by jordisala1991** [comment]
I think we should removed `Changed` and `Added` section of changelog. Unless we manage to rephrase `Added` so it explains that this new class will simplify development of commands that use dialogs.

**Comment by sinoptik** [comment]
@greg0ire I understood you. I think that @jordisala1991  is right. Just removed `Changed` and `Added` section and describe all changes in  `Fixed` part.

**Comment by greg0ire** [comment]
@sinoptik way better! Good job!

**Comment by greg0ire** [comment]
Thanks @sinoptik !

**Pull request merged**